### PR TITLE
chore: Upgrade RTS Dependencies

### PR DESCRIPTION
**Pull request by luc-blaeser**
Upgrading Motoko RTS dependencies:
* LLVM 18.1.8
* Rust 1.82.0

Special aspects:
* Support new `dylink.0` instead of `dylink`.
* Support `GOT.mem` for RTS in the Motoko linker.
* Adjusting the RTS code in line with the latest Rust features/rules.
* No longer needing `emscripten`: Use custom Rust targets .
* Get rid of `emscripten` by using custom Rust targets for generating PIC-relocatable Wasm.
* Get rid of `musl` by using Rust builtins instead.
* Float formatting changes due to removal of `musl`: The output format of `NaN` and `infinity` has changed. Moreover, one particular output format (hexadecimal float) is no longer supported. Updating the documentation and base library is needed.
* Optimization: Use bulk-memory operations because memcpy of Rust is slower than `musl`.

Note:
* If your `nix-shell` crashes with a segmentation fault with this PR, you need to upgrade `nix`. For this, you need to manually uninstall and re-install `nix`.

### CODE REVIEW

File: rts/Makefile [hunk 1]
```diff
@@ -136,9 +140,20 @@ _build/wasm32:
...
+	-mrelocation-model pic -pic-level 1 -mframe-pointer=none \
+	-ffp-contract=on -fno-rounding-math -mconstructor-aliases \
+	-target-feature +mutable-globals -fvisibility=hidden \
+	-Werror=implicit-function-declaration \
+	-iwithsysroot/include/fakesdl -iwithsysroot/include/compat \
+	-mllvm -combiner-global-alias-analysis=false \
+	-mllvm -enable-emscripten-sjlj -mllvm -disable-lsr \
+	-Wno-override-module
+RTS_LL_64_DEBUG=$(RTS_LL_64)
+RTS_LL_64_RELEASE=$(RTS_LL_64) -Oz
```
Comment (crusso) [inline rts/Makefile:156]: Could it be that the change from 02 to 0z is causing the rts size increase?

Comment (luc-blaeser) [reply rts/Makefile:156]: I believe the increase would be even higher with `O2`. I'll double check and let you know.

Comment (luc-blaeser) [reply rts/Makefile]: I removed `emscripten` by dedicated Rust targets that generate PIC shared Wasm libraries.

**Comment by luc-blaeser** [comment]
> Comparing from [247aa05](https://github.com/dfinity/motoko/commit/247aa056afe9d5499f9f79d78d6d38117b8478ae) to [a6d1305](https://github.com/dfinity/motoko/commit/a6d1305b087e5968ddba40452cacaa0ef1223a80): In terms of gas, 4 tests regressed, 1 tests improved and the mean change is -0.0%. In terms of size, 5 tests regressed and the mean change is +29.3%.

Now optimized the both size and performance. Would be nice if we could rerun this benchmark.

**Comment by luc-blaeser** [comment]
Size compression is no longer needed for the specific Rust target backends.
Instead, we can again optimize for speed:

This is the latest perf diff on CI (locally repeated):
```
Comparing from b7fa63936cf708952deae0dd0b0429faa1baa119 to a10ab53a1ec3e5e48dbce36d6c7dd8547503c85c:
In terms of gas, 3 tests regressed, 2 tests improved and the mean change is -0.0%.
In terms of size, 5 tests improved and the mean change is -1.6%.
```

**Comment by luc-blaeser** [comment]
According to the GC benchmark, it seems that this upgrade/cleanup even improves performance:

Mode | Before | Now | Improvement
-------------------|--------|-----|-----
Classical (copying GC) | 2.33e10 | 2.29e10 | 1.7%
Classical (incremental GC) | 2.39e10 | 2.35e10 | 1.7%
EOP | 2.54e0 | 2.50e10 | 1.6%

File: nix/sources.json [hunk 2]
```diff
@@ -70,15 +70,15 @@
         "version": "v1.2.0"
     },
     "motoko-base": {
-        "branch": "next-moc",
```
Comment (luc-blaeser) [inline nix/sources.json:73]: This is to be changed back to `next-moc` once https://github.com/dfinity/motoko-base/pull/589 is merged.

File: nix/default.nix [hunk 3]
```diff
@@ -78,7 +78,7 @@ let
                 };
               };
 
-              # No testing of atdgen, as it pulls in python stuff, tricky on musl
+              # No testing of atdgen, as it pulls in python stuff
               atdgen = super.ocamlPackages.atdgen.overrideAttrs { doCheck = false; };
```
Comment (ggreif) [inline nix/default.nix:82]: this could be `doCheck = true` now?

Comment (ggreif) [reply nix/default.nix:82]: or rather uncustomized?

File: nix/default.nix [hunk 4]
```diff
@@ -91,13 +91,11 @@ let
 
         # Rust nightly
         (self: super: let
-          rust-channel = self.moz_overlay.rustChannelOf { date = "2023-04-21"; channel = "nightly"; };
+          rust-channel = self.moz_overlay.rustChannelOf { date = "2024-07-28"; channel = "nightly"; };
         in rec {
           rustc-nightly = rust-channel.rust.override {
             targets = [
-               "wasm32-unknown-emscripten"
```
Comment (ggreif) [inline nix/default.nix:98]: oh, why were these here?

File: rts/motoko-rts/Cargo.toml [hunk 5]
```diff
@@ -44,4 +44,3 @@ panic = "abort"
 [profile.release]
 panic = "abort"
 codegen-units = 1 # makes debugging easier
-opt-level = "s" # optimize for size
```
Comment (ggreif) [inline rts/motoko-rts/Cargo.toml:47]: Why this change?

File: rts/motoko-rts/src/float.rs [hunk 6]
```diff
@@ -1,7 +1,40 @@
...
+use crate::memory::Memory;
+use crate::text::text_of_ptr_size;
+use crate::types::{Bytes, Value};
 
-#[classical_persistence]
-pub mod classical;
+use motoko_rts_macros::ic_mem_fn;
 
-use motoko_rts_macros::{classical_persistence, enhanced_orthogonal_persistence};
+// The meaning of the `mode` parameter is documented in motoko-base, function Float.format()
```
Comment (ggreif) [inline rts/motoko-rts/src/float.rs:7]: nice move!

File: src/linking/linkModule.ml [hunk 7]
```diff
@@ -715,26 +742,22 @@ let find_fun_export (name : name) (exports : export list) : var option =
...
-      else
-        match globals with
-        | [] -> List.map (fun (_, global) -> global @@ no_region) insert
-        | global :: globals -> global :: go (Int32.add current_idx 1l) globals rest
-  in
-  go 0l globals0 insert0
+  let got_mem_str = Lib.Utf8.decode "GOT.mem" in
+  let is_got name = name = got_func_str || name = got_mem_str in
+  List.filter (fun import -> not (is_got import.it.module_name)) imports
+
```
Comment (ggreif) [inline src/linking/linkModule.ml:760]:
```suggestion
  List.filter (fun import -> not (is_got import.it.module_name)) imports
```

File: src/wasm-exts/customModuleEncode.ml [hunk 8]
```diff
@@ -915,7 +916,7 @@ let encode (em : extended_module) =
       patch s (p + 1) (lsb (n lsr 8));
       patch s (p + 2) (lsb (n lsr 16));
       patch s (p + 3) (lsb (n lsr 24))
-    let dw_patches = ref Fun.id
+    let dw_patches = ref (fun i -> i)
```
Comment (ggreif) [inline src/wasm-exts/customModuleEncode.ml:919]: heh, why this?

File: test/run/float-fmt.mo [hunk 9]
```diff
@@ -30,11 +29,5 @@ Prim.debugPrint(Prim.floatToFormattedText(pi, 16, 2));
...
 
-Prim.debugPrint "hex Float";
-Prim.debugPrint(Prim.floatToFormattedText(pi, 0, 3));
-Prim.debugPrint(Prim.floatToFormattedText(pi, 1, 3));
-Prim.debugPrint(Prim.floatToFormattedText(pi, 4, 3));
-Prim.debugPrint(Prim.floatToFormattedText(pi, 9, 3));
-Prim.debugPrint(Prim.floatToFormattedText(pi, 16, 3));
-Prim.debugPrint(Prim.floatToFormattedText(pi, 17, 3));
-Prim.debugPrint(Prim.floatToFormattedText(pi, 20, 3));
+// Hex Float is no longer supported
```
Comment (ggreif) [inline test/run/float-fmt.mo:32]: did we ever deprecate them?

**Review by ggreif — DISMISSED** [review]
LGTM! Heroic work on the `dylink0` stuff. Being stuck on `clang-13` gave me nightmares...
Nice to see the cycle improvements too. Getting rid of `musl` means one less wheel to maintain.

Good work! Added a few Qs...

File: rts/motoko-rts/Cargo.toml (hunk 5 above)
Comment (ggreif) [reply rts/motoko-rts/Cargo.toml:47]: Okay, explained in https://github.com/dfinity/motoko/pull/4677#issuecomment-2402465457

File: src/linking/linkModule.ml [hunk 10]
```diff
@@ -743,75 +766,206 @@ let mk_i32_global (i : int32) =
...
+  in
 
-   `tbe_size` is the size of the table in the merged module before adding
-   GOT.func functions. *)
-let replace_got_func_imports (tbl_size : int32) (imports : (int32 * int32) list) (m : module_') : module_' =
-  (* null check to avoid adding empty elem section *)
-  if imports = [] then
-    m
+  (* Computes the absolute address of the GOT.mem data pointer *)
+  let data_pointer exported_global_index = if uses_memory64 m then
```
Comment (crusso) [inline src/linking/linkModule.ml:868]: Not that you should do this now, but I wonder if it would be possible to just bump the (mutable?) globals by the offset using an initialization function, rather than recomputing the absolute address at each read. Another advantage, apart from perf, would be that any debug info would remain valid (rather than disturbing the instruction offset in the code.

File: src/linking/linkModule.ml [hunk 11]
```diff
@@ -54,41 +54,57 @@ importing module would work:
...
+  module section.
+
+  For each `GOT.func`, we create an element in the table with the function index
+  determined in the first step. The `GOT.func` import is replaced by a global
+  that refers to the new table element.
+
+  For each `GOT.mem`, we replace the GOT accesses in the AST by code that computes 
+  the absolute address of the data. This is the library's memory base plus the 
+  relative data offset stored in the global as determined in the first step.  
+  The `GOT.mem` import is replaced by a dummy global that only serves to maintain
```
Comment (crusso) [inline src/linking/linkModule.ml:94]: I wonder if the dummy global could instead hold the absolute/resolved address, and we replace references to the import by references to the resolved global? But I don't fully understand this stuff..

**Review by crusso — DISMISSED** [review]
Great work! Not sure I fully understand the intricacies so 🤞

File: nix/default.nix (hunk 4 above)
Comment (luc-blaeser) [reply nix/default.nix:98]: We used emscripten to get PIC-relocatable shared Wasm modules for the RTS. This is now redundant by the customized RTS backends.

File: default.nix [hunk 12]
```diff
@@ -35,35 +35,31 @@ in
 
 let
   rtsBuildInputs = with nixpkgs; [
-    # pulls in clang (wrapped) and clang-13 (unwrapped)
-    llvmPackages_13.clang
-    # pulls in wasm-ld
-    llvmPackages_13.lld
```
[the remark this answers is not in the document]
Comment (luc-blaeser) [reply default.nix:41]: Yes, `wasm-ld` is part of `bintools`.

File: rts/motoko-rts/src/float.rs (hunk 6 above)
Comment (luc-blaeser) [reply rts/motoko-rts/src/float.rs:7]: Thank you!

File: rts/motoko-rts/src/allocator.rs [hunk 13]
```diff
@@ -41,33 +40,3 @@ unsafe impl GlobalAlloc for EphemeralAllocator {
 
 #[global_allocator]
 static ALLOCATOR: EphemeralAllocator = EphemeralAllocator;
-
```
[the remark this answers is not in the document]
Comment (luc-blaeser) [reply rts/motoko-rts/src/allocator.rs:44]: I believe this is due to the new Rust version.

File: src/wasm-exts/customModuleEncode.ml (hunk 8 above)
Comment (luc-blaeser) [reply src/wasm-exts/customModuleEncode.ml:919]: Not needed, accidental change...

File: src/linking/linkModule.ml (hunk 11 above)
Comment (luc-blaeser) [reply src/linking/linkModule.ml:94]: I find this actually a good idea. If okay, I'll put it on backlog as GOT.mem are seldom.

File: src/linking/linkModule.ml (hunk 7 above)
Comment (luc-blaeser) [reply src/linking/linkModule.ml:760]: I'll clean up in one commit.

File: src/linking/linkModule.ml (hunk 10 above)
Comment (luc-blaeser) [reply src/linking/linkModule.ml:868]: A good idea. If okay, I'll remember for later exploration, as GOT.mem are relatively seldom.

File: src/wasm-exts/customModuleEncode.ml (hunk 8 above)
Comment (luc-blaeser) [reply src/wasm-exts/customModuleEncode.ml:919]: Adjusted

File: nix/default.nix (hunk 3 above)
Comment (luc-blaeser) [reply nix/default.nix:82]: Done

**Review by crusso — APPROVED** [review]

**Review by crusso — APPROVED** [review]
LGTM

File: nix/sources.json [hunk 14]
```diff
@@ -70,15 +70,15 @@
...
         "owner": "dfinity",
         "repo": "motoko-base",
-        "rev": "21a80ab250501a5de3365114abb242559573fecc",
-        "sha256": "0zm60azafdfczdiqkqs1fwgxf3crb2bk0vs7y09rvnwkqgqkxxw8",
+        "rev": "1bece926bfc417db5661c66d8617f6b09ab06b95",
+        "sha256": "02i8s4ybjmbbf72jclpvqcsxh4dypgsxqsagiga64jqyprmqzvlr",
         "type": "tarball",
-        "url": "https://github.com/dfinity/motoko-base/archive/21a80ab250501a5de3365114abb242559573fecc.tar.gz",
+        "url": "https://github.com/dfinity/motoko-base/archive/1bece926bfc417db5661c66d8617f6b09ab06b95.tar.gz",
         "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
```
[the remark this answers is not in the document]
Comment (luc-blaeser) [reply nix/sources.json]: Updated

**Comment by luc-blaeser** [comment]
Thank you for reviewing, Claudio and Gabor!

**Review by crusso — DISMISSED** [review]

**Review by crusso — APPROVED** [review]

**Review by ggreif — APPROVED** [review]
LGTM!
